crates/matrix-sdk/src/encryption/recovery/mod.rs

```diff
@@ -0,0 +1,526 @@
+        self.client.account().set_account_data(SecretStorageDisabledContent {}).await?;
+        self.client.account().set_account_data(BackupDisabledContent { disabled: true }).await?;
+        self.update_recovery_state().await?;
+        // TODO: Do we want to "delete" the known secrets as well?
+
+        Ok(())
+    }
+
+    /// Reset the recovery key.
+    ///
+    /// This will rotate the secret storage key and re-upload all the secrets to
+    /// the [`SecretStore`].
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// let new_recovery_key =
+    ///     recovery.reset_key().with_passphrase("my passphrase").await;
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub fn reset_key(&self) -> Reset<'_> {
+        // TODO: Should this only be possible if we're in the RecoveryState::Enabled
+        // state? Otherwise we'll create a new secret store but won't be able to
+        // upload all the secrets.
+        Reset::new(self)
+    }
+
+    /// Reset the recovery key but first import all the secrets from secret
+    /// storage.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
+    ///
+    /// let recovery = client.encryption().recovery();
+    ///
+    /// let new_recovery_key = recovery
+    ///     .recover_and_reset("my old passphrase or key")
+    ///     .with_passphrase("my new passphrase")
+    ///     .await?;
+    /// # anyhow::Ok(()) };
+    /// ```
+    pub fn recover_and_reset<'a>(&'a self, old_key: &'a str) -> RecoverAndReset<'_> {
+        RecoverAndReset::new(self, old_key)
+    }
+
+    /// Recover all the secrets from the homeserver.
+    ///
+    /// This method is a convenience method around the
+    /// [`SecretStore::import_secrets()`] method, please read the documentation
+    /// of this method for more information about what happens if you call
+    /// this method.
+    ///
+    /// In short, this method will turn a newly created [`Client`] into a fully
+    /// end-to-end encryption enabled client.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # use matrix_sdk::{Client, encryption::recovery::RecoveryState};
+    /// # use url::Url;
+    /// # async {
+    /// # let homeserver = Url::parse("http://example.com")?;
+    /// # let client = Client::new(homeserver).await?;
+    /// use futures_util::StreamExt;
```

Comment:
… and here.

Reply:
Same 77444856aab8b321070293c3d661a4efa2f33082